### PR TITLE
Rapidjson parity

### PR DESCRIPTION
**Pull request by ccaraman**

### CODE REVIEW

File: source/common/json/json_loader.cc [hunk 1]
```diff
@@ -1,125 +1,118 @@
...
 
 FileLoader::FileLoader(const std::string& file_path) {
-  json_error_t error;
-  json_ = json_load_file(file_path.c_str(), 0, &error);
-  if (!json_) {
-    throw Exception(fmt::format("json parsing error: {} line: {} column: {}", error.text,
-                                error.line, error.column));
+  char buffer[4096];
+
+  FILE* fp = fopen(file_path.c_str(), "r");
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: You are leaking here. Can you use C++ iostream?

File: source/common/json/json_loader.cc [hunk 2]
```diff
@@ -1,125 +1,118 @@
...
+  if (!fp) {
+    throw Exception(fmt::format("File:{} doesn't exist.", file_path));
+  }
+
+  rapidjson::FileReadStream fs(fp, buffer, sizeof(buffer));
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: what is buffer used for here? Why does it have to be specified? Why 4K? Seems odd there is no API to just deal with buffering internally.

File: source/common/json/json_loader.cc [hunk 3]
```diff
@@ -1,125 +1,118 @@
...
+  if (document_.ParseStream(fs).HasParseError()) {
+    throw Exception(fmt::format("Error(offset {}): {}\n",
+                                static_cast<unsigned>(document_.GetErrorOffset()),
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: What is offset? Does it include line number?

File: source/exe/CMakeLists.txt [hunk 4]
```diff
@@ -20,6 +20,7 @@ target_link_libraries(envoy rt)
 target_link_libraries(envoy dl)
 
 include_directories(SYSTEM ${ENVOY_OPENSSL_INCLUDE_DIR})
+include_directories(SYSTEM ${ENVOY_RAPIDJSON_INCLUDE_DIR})
```
Comment (mattklein123) [inline source/exe/CMakeLists.txt]: why is this needed?

File: source/server/CMakeLists.txt [hunk 5]
```diff
@@ -24,6 +24,7 @@ include_directories(SYSTEM ${ENVOY_HTTP_PARSER_INCLUDE_DIR})
 include_directories(${ENVOY_NGHTTP2_INCLUDE_DIR})
 include_directories(SYSTEM ${ENVOY_OPENSSL_INCLUDE_DIR})
 include_directories(SYSTEM ${ENVOY_LIGHTSTEP_TRACER_INCLUDE_DIR})
+include_directories(SYSTEM ${ENVOY_RAPIDJSON_INCLUDE_DIR})
```
Comment (mattklein123) [inline source/server/CMakeLists.txt]: why is this needed?

File: source/common/json/json_loader.h [hunk 6]
```diff
@@ -25,7 +28,12 @@ typedef std::function<bool(const std::string&, const Object&)> ObjectCallback;
  */
 class Object {
 public:
-  Object(json_t* json, const std::string& name) : json_(json), name_(name) {}
+  Object(const rapidjson::Value& value, const std::string& name) : name_(name), document_() {
+    document_.CopyFrom(value, document_.GetAllocator());
+  }
+
+  // Empty Object
+  Object(const std::string& name) : name_(name), document_() {}
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: default initializer not needed

File: source/common/json/json_loader.h [hunk 7]
```diff
@@ -25,7 +28,12 @@ typedef std::function<bool(const std::string&, const Object&)> ObjectCallback;
  */
 class Object {
 public:
-  Object(json_t* json, const std::string& name) : json_(json), name_(name) {}
+  Object(const rapidjson::Value& value, const std::string& name) : name_(name), document_() {
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: default initializer not needed

File: source/common/json/json_loader.h [hunk 8]
```diff
@@ -4,7 +4,10 @@
 
 #include "common/common/non_copyable.h"
 
-struct json_t;
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wold-style-cast"
+#include "rapidjson/document.h"
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: Do not leak this in header. Find a way around this even if you need to expose an interface.

File: source/common/json/json_loader.cc [hunk 9]
```diff
@@ -129,41 +122,36 @@ std::string Object::getString(const std::string& name, const std::string& defaul
...
+    string_array.push_back(object.document_.GetString());
   }
 
   return string_array;
 }
 
 double Object::getDouble(const std::string& name) const {
-  json_t* real = json_object_get(json_, name.c_str());
-  if (!real || !json_is_real(real)) {
+  if (!document_.HasMember(name.c_str()) || !document_[name.c_str()].IsDouble()) {
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: can you avoid multiple hash lookups here and elsewhere

File: source/common/json/json_loader.cc (hunk 1 above)
Comment (RomanDzhabarov) [inline source/common/json/json_loader.cc]: if not, you can use CSmartPtr and supply fclose not to leak fd.

File: source/common/json/json_loader.cc [hunk 10]
```diff
@@ -1,125 +1,118 @@
 #include "json_loader.h"
 
-// Do not let jansson leak outside of this file.
+// Do not let RapidJson leak outside of this file.
 #pragma GCC diagnostic push
 #pragma GCC diagnostic ignored "-Wold-style-cast"
-#include "jansson.h"
+#include <rapidjson/error/en.h>
```
Comment (RomanDzhabarov) [inline source/common/json/json_loader.cc]: minor: s/</"

File: source/common/json/json_loader.cc (hunk 3 above)
Comment (RomanDzhabarov) [inline source/common/json/json_loader.cc]: wondering, why casting to unsigned_int?

File: source/server/CMakeLists.txt (hunk 5 above)
Comment (ccaraman) [inline source/server/CMakeLists.txt]: It made the compiler happy, back when I was playing with it. I don't think this will be necessary once I find a way to get rid of the refernce to rapidjson/document.h in the json_loader.h

File: source/common/json/json_loader.cc (hunk 2 above)
Comment (ccaraman) [inline source/common/json/json_loader.cc]: Removed. Switched to using iostream.

File: source/common/json/json_loader.cc (hunk 10 above)
Comment (ccaraman) [inline source/common/json/json_loader.cc]: Fixed.

File: source/common/json/json_loader.cc (hunk 9 above)
Comment (ccaraman) [inline source/common/json/json_loader.cc]: Removed multiple hash lookups.

File: source/exe/CMakeLists.txt (hunk 4 above)
Comment (ccaraman) [inline source/exe/CMakeLists.txt]: Removed. It was due to the initial integration of rapidjson. No longer needed now that rapidjson headers are only in the .cc file.

File: source/common/json/json_loader.cc [hunk 11]
```diff
@@ -1,178 +1,194 @@
...
-  if (!json_is_array(json_)) {
-    throw Exception(fmt::format("'{}' is not an array", name_));
+    return object_array;
   }
 
-  std::vector<Object> object_array;
-  for (size_t i = 0; i < json_array_size(json_); i++) {
-    object_array.emplace_back(json_array_get(json_, i), name_ + " (array item)");
+  bool getBoolean(const std::string& name) const override {
+    if (!document_[name.c_str()].IsBool()) {
```
Comment (RomanDzhabarov) [inline source/common/json/json_loader.cc]: not to familiar with api, but can we
1) auto &blah = document_[name.c_str()];
and query internal map or something once.

if (!blah.IsBool()) {
}

return blah.GetBool();

File: include/envoy/json/json_object.h [hunk 12]
```diff
@@ -0,0 +1,133 @@
...
+   * @param default_value supplies the value to return if name does not exist.
+   * @return int64_t the value.
+   */
+  virtual int64_t getInteger(const std::string& name, int64_t default_value) const PURE;
+
+  /**
+   * Get a sub-object by name.
+   * @param name supplies the key name.
+   * @param allow_empty supplies whether to return an empty object if the key does not exist.
+   * @return Object the sub-object.
```
Comment (mattklein123) [inline include/envoy/json/json_object.h]: wrong type

File: include/envoy/json/json_object.h [hunk 13]
```diff
@@ -0,0 +1,133 @@
...
+   */
+  virtual AbstractObjectPtr getObject(const std::string& name, bool allow_empty = false) const PURE;
+
+  /**
+   * Get an array by name.
+   * @param name supplies the key name.
+   * @return std::vector<Object> the array of JSON objects.
```
Comment (mattklein123) [inline include/envoy/json/json_object.h]: wrong type

File: include/envoy/json/json_object.h [hunk 14]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Get a double value by name.
+   * @param name supplies the key name.
+   * @param default_value supplies the value to return if name does not exist.
+   * @return double the value.
+   */
+  virtual double getDouble(const std::string& name, double default_value) const PURE;
+
+  /**
+   * Iterate Object and call callback on key-value pairs
```
Comment (mattklein123) [inline include/envoy/json/json_object.h]: wrong type / please fix grammar

File: include/envoy/json/json_object.h [hunk 15]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * @return TRUE if the object contains an element.
+   * @param name supplies the key name to lookup.
+   */
+  virtual bool hasObject(const std::string& name) const PURE;
+
+  /**
+   * @return a vector of the AbstractObject that are members of the currect AbstractObject.
+   */
+  virtual std::vector<AbstractObjectPtr> getMembers() const PURE;
```
Comment (mattklein123) [inline include/envoy/json/json_object.h]: can't this be implemented in terms of iterate? Why is this needed?

Comment (mattklein123) [inline include/envoy/json/json_object.h]: Or get rid of iterate() from the public interface

File: source/common/json/json_loader.cc [hunk 16]
```diff
@@ -1,178 +1,194 @@
...
-  if (!json_) {
-    throw Exception(fmt::format("json parsing error: {} line: {} column: {}", error.text,
-                                error.line, error.column));
+/**
+ * Implementation of AbstractObject.
+ */
+class AbstractObjectImpl : public AbstractObject {
+public:
+  AbstractObjectImpl(const rapidjson::Document& document, const std::string& name) : name_(name) {
+    document_.CopyFrom(document, document_.GetAllocator());
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: We should not be copying anything. If needed a reference to the parent document should be used.

File: source/common/json/json_loader.cc [hunk 17]
```diff
@@ -1,178 +1,194 @@
...
-
-FileLoader::~FileLoader() { json_decref(json_); }
 
-StringLoader::StringLoader(const std::string& json) {
-  json_error_t error;
-  json_ = json_loads(json.c_str(), 0, &error);
-  if (!json_) {
-    throw Exception(fmt::format("json parsing error: {}", error.text));
+  AbstractObjectImpl(const rapidjson::Value& value, const std::string& name) : name_(name) {
+    document_.CopyFrom(value, document_.GetAllocator());
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: no copying

File: test/common/json/json_loader_test.cc [hunk 18]
```diff
@@ -122,11 +124,6 @@ TEST(JsonLoaderTest, Integer) {
     EXPECT_EQ(std::numeric_limits<int64_t>::max(), json.getInteger("max"));
     EXPECT_EQ(std::numeric_limits<int64_t>::min(), json.getInteger("min"));
   }
-
-  {
```
Comment (mattklein123) [inline test/common/json/json_loader_test.cc]: what was this test testing and why is it no longer failing

File: source/common/json/json_loader.h [hunk 19]
```diff
@@ -16,150 +28,103 @@ class Exception : public EnvoyException {
...
 };
 
 /**
  * Loads JSON from a string.
  */
 class StringLoader : NonCopyable, public Object {
 public:
-  StringLoader(const std::string& json);
-  ~StringLoader();
+  StringLoader(const std::string& json) { abstract_object_ = Factory::LoadFromString(json); }
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: initializer list

File: source/common/json/json_loader.h [hunk 20]
```diff
@@ -16,150 +28,103 @@ class Exception : public EnvoyException {
...
+  }
+
+  Object getObject(const std::string& name, bool allow_empty = false) const {
+    AbstractObjectPtr abstract_object = abstract_object_->getObject(name, allow_empty);
+    return Object(std::move(abstract_object));
+  }
+
+  std::vector<Object> getObjectArray(const std::string& name) const {
+    std::vector<AbstractObjectPtr> abstract_object_array = abstract_object_->getObjectArray(name);
+    std::vector<Object> return_vector;
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: reserve capacity ahead of time

File: source/common/json/json_loader.h [hunk 21]
```diff
@@ -16,150 +28,103 @@ class Exception : public EnvoyException {
...
+  }
+
+  int64_t getInteger(const std::string& name) const { return abstract_object_->getInteger(name); }
+
+  int64_t getInteger(const std::string& name, int64_t default_value) const {
+    return abstract_object_->getInteger(name, default_value);
+  }
+
+  Object getObject(const std::string& name, bool allow_empty = false) const {
+    AbstractObjectPtr abstract_object = abstract_object_->getObject(name, allow_empty);
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: local var not needed

File: source/common/json/json_loader.h [hunk 22]
```diff
@@ -16,150 +28,103 @@ class Exception : public EnvoyException {
...
-  bool hasObject(const std::string& name) const;
+  Object() {}
+
+  Object(AbstractObjectPtr&& abstract_object_ptr) {
+    abstract_object_ = std::move(abstract_object_ptr);
+  }
+
+  std::vector<Object> asObjectArray() const {
+    std::vector<AbstractObjectPtr> abstract_object_array = abstract_object_->asObjectArray();
+    std::vector<Object> return_vector;
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: reserve

File: source/common/json/json_loader.h [hunk 23]
```diff
@@ -16,150 +28,103 @@ class Exception : public EnvoyException {
...
-
-  /**
-   * @return TRUE if the object contains an element.
-   * @param name supplies the key name to lookup.
-   */
-  bool hasObject(const std::string& name) const;
+  Object() {}
+
+  Object(AbstractObjectPtr&& abstract_object_ptr) {
+    abstract_object_ = std::move(abstract_object_ptr);
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: initializer list

File: source/common/json/json_loader.cc [hunk 24]
```diff
@@ -1,178 +1,194 @@
...
-void Object::iterate(const ObjectCallback& callback) {
-  const char* key;
-  json_t* value;
+  const std::string getName() const override { return name_; }
 
-  json_object_foreach(json_, key, value) {
-    Object object_value(value, "root");
-    std::string object_key(key);
+private:
+  std::string name_;
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: const

File: source/common/json/json_loader.cc [hunk 25]
```diff
@@ -1,178 +1,194 @@
...
-}
-
-double Object::getDouble(const std::string& name, double default_value) const {
-  if (!json_object_get(json_, name.c_str())) {
-    return default_value;
-  } else {
-    return getDouble(name);
+  std::vector<AbstractObjectPtr> getMembers() const override {
+    std::vector<AbstractObjectPtr> return_vector;
+    for (rapidjson::Value::ConstMemberIterator itr = document_.MemberBegin();
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: if this is how getMembers() is implemented I don't understand why we need a new getMembers() method at all. Just use this to implement iterate.

File: test/common/json/json_loader_test.cc (hunk 18 above)
Comment (ccaraman) [inline test/common/json/json_loader_test.cc]: This was testing Max and Min Integer. Unfortunately, rapidjson doesn't throw an exception during Parsing. We should push this validation to the schema. :(

File: include/envoy/json/json_object.h [hunk 26]
```diff
@@ -0,0 +1,133 @@
...
+  /**
+   * Get a boolean value by name.
+   * @param name supplies the key name.
+   * @return bool the value.
+   */
+  virtual bool getBoolean(const std::string& name) const PURE;
+
+  /**
+   * Get a boolean value by name.
+   * @param name supplies the  key name.
```
Comment (ccaraman) [inline include/envoy/json/json_object.h]: fixed.

File: include/envoy/json/json_object.h (hunk 13 above)
Comment (ccaraman) [inline include/envoy/json/json_object.h]: Fixed.

File: include/envoy/json/json_object.h (hunk 15 above)
Comment (ccaraman) [inline include/envoy/json/json_object.h]: Removed.

File: source/common/json/json_loader.cc [hunk 27]
```diff
@@ -1,178 +1,194 @@
...
-Object::EmptyObject Object::empty_;
-
-Object::EmptyObject::EmptyObject() : json_(json_object()) {}
+  std::vector<AbstractObjectPtr> asObjectArray() const override {
+    if (!document_.IsArray()) {
+      throw Exception(fmt::format("'{}' is not an array", name_));
+    }
 
-Object::EmptyObject::~EmptyObject() { json_decref(json_); }
+    std::vector<AbstractObjectPtr> object_array;
```
Comment (ccaraman) [inline source/common/json/json_loader.cc]: Fixed here and in json_loader.h

File: source/common/json/json_loader.h (hunk 23 above)
Comment (ccaraman) [inline source/common/json/json_loader.h]: Fixed.

File: source/common/json/json_loader.h (hunk 22 above)
Comment (ccaraman) [inline source/common/json/json_loader.h]: Fixed.

File: source/common/json/json_loader.h (hunk 19 above)
Comment (ccaraman) [inline source/common/json/json_loader.h]: Fixed.

File: source/common/json/json_loader.h (hunk 20 above)
Comment (ccaraman) [inline source/common/json/json_loader.h]: Reserved

File: source/common/json/json_loader.h (hunk 21 above)
Comment (ccaraman) [inline source/common/json/json_loader.h]: removed extra local var.

File: docs/install/requirements.rst [hunk 28]
```diff
@@ -14,10 +14,10 @@ Envoy has the following requirements:
 * `libevent <http://libevent.org/>`_ (last tested with 2.0.22)
 * `tclap <http://tclap.sourceforge.net/>`_ (last tested with 1.2.1)
 * `gperftools <https://github.com/gperftools/gperftools>`_ (last tested with 2.5.0)
-* `jansson <https://github.com/akheron/jansson>`_ (last tesed with 2.7)
-* `openssl <https://www.openssl.org/>`_ (last tesed with 1.0.2i)
+* `openssl <https://www.openssl.org/>`_ (last tested with 1.0.2i)
```
Comment (ccaraman) [inline docs/install/requirements.rst]: I will remove everything related to jansson with the pr for Jira 615

File: source/common/json/json_loader.cc [hunk 29]
```diff
@@ -1,178 +1,200 @@
...
+    if (!value_.IsArray()) {
+      throw Exception(fmt::format("'{}' is not an array", name_));
+    }
 
-Object::EmptyObject::EmptyObject() : json_(json_object()) {}
+    std::vector<AbstractObjectPtr> object_array;
+    object_array.reserve(value_.Size());
 
-Object::EmptyObject::~EmptyObject() { json_decref(json_); }
+    for (rapidjson::SizeType i = 0; i < value_.Size(); i++) {
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: range based for loop?

File: source/common/json/json_loader.cc [hunk 30]
```diff
@@ -1,178 +1,200 @@
...
-  if (!json_is_array(json_)) {
-    throw Exception(fmt::format("'{}' is not an array", name_));
+    return object_array;
   }
 
-  std::vector<Object> object_array;
-  for (size_t i = 0; i < json_array_size(json_); i++) {
-    object_array.emplace_back(json_array_get(json_, i), name_ + " (array item)");
+  bool getBoolean(const std::string& name) const override {
+    const rapidjson::Value& value = value_[name.c_str()];
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: what happens here if name.c_str() does not exist?

File: source/common/json/json_loader.cc [hunk 31]
```diff
@@ -1,178 +1,200 @@
...
-}
+    std::vector<AbstractObjectPtr> object_array;
+    object_array.reserve(value.Size());
 
-Object Object::getObject(const std::string& name, bool allow_empty) const {
-  json_t* object = json_object_get(json_, name.c_str());
-  if (!object && allow_empty) {
-    object = empty_.json_;
-  }
+    for (rapidjson::SizeType i = 0; i < value.Size(); i++) {
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: range for loop

File: source/common/json/json_loader.cc [hunk 32]
```diff
@@ -1,178 +1,200 @@
...
 
-  return string_array;
-}
-
-double Object::getDouble(const std::string& name) const {
-  json_t* real = json_object_get(json_, name.c_str());
-  if (!real || !json_is_real(real)) {
-    throw Exception(fmt::format("key '{}' missing or not a double in '{}'", name, name_));
+  void iterate(const AbstractObjectCallback& callback) override {
+    for (rapidjson::Value::ConstMemberIterator itr = value_.MemberBegin();
```
Comment (mattklein123) [inline source/common/json/json_loader.cc]: range for loop

File: source/common/json/json_loader.cc (hunk 29 above)
Comment (ccaraman) [inline source/common/json/json_loader.cc]: Shoot that was the one I forgot to fix. Fixing now...  And will fix the rest of the for loops.

File: source/common/json/json_loader.cc (hunk 30 above)
Comment (ccaraman) [inline source/common/json/json_loader.cc]: Changed it to use FindMember.

File: source/common/json/json_loader.cc [hunk 33]
```diff
@@ -1,178 +1,200 @@
...
+      return default_value;
+    } else {
+      return getBoolean(name);
+    }
   }
 
-  return json_boolean_value(boolean);
-}
+  int64_t getInteger(const std::string& name) const override {
+    const rapidjson::Value& value = value_[name.c_str()];
```
Comment (ccaraman) [inline source/common/json/json_loader.cc]: changed to use FindMember

File: source/common/json/json_loader.cc (hunk 32 above)
Comment (ccaraman) [inline source/common/json/json_loader.cc]: changed to range for loop

File: source/common/json/json_loader.h [hunk 34]
```diff
@@ -1,12 +1,30 @@
...
 
 #include "envoy/common/exception.h"
+#include "envoy/common/pure.h"
 
 #include "common/common/non_copyable.h"
 
-struct json_t;
-
 namespace Json {
+class Object;
```
Comment (mattklein123) [inline source/common/json/json_loader.h]: split object out into public include and only define the factory stuff in this file.

Comment (ccaraman) [inline source/common/json/json_loader.h]: split.

File: test/common/json/json_loader_test.cc [hunk 35]
```diff
@@ -110,39 +110,35 @@ TEST(JsonLoaderTest, Basic) {
...
 
 TEST(JsonLoaderTest, Integer) {
   {
-    StringLoader json("{\"max\":9223372036854775807, \"min\":-9223372036854775808}");
-    EXPECT_EQ(std::numeric_limits<int64_t>::max(), json.getInteger("max"));
-    EXPECT_EQ(std::numeric_limits<int64_t>::min(), json.getInteger("min"));
-  }
-
-  {
-    EXPECT_THROW(StringLoader("{\"val\":9223372036854775808}"), Exception);
```
Comment (ccaraman) [inline test/common/json/json_loader_test.cc]: Readded.
